go/gilded-rose.go

```diff
@@ -1,57 +1,139 @@
 package main
 
+// Item represents an item for sale at the inn
 type Item struct {
 	name            string
 	sellIn, quality int
 }
 
-func UpdateQuality(items []*Item) {
-	for i := 0; i < len(items); i++ {
+// simple func to return a min of 2 integers
+// useful in this context to stop items value going over 50
+func min(x int, y int) int {
+	if x == y {
+		return x
+	}
+	if x < y {
+		return x
+	}
+	return y
+}
 
-		if items[i].name != "Aged Brie" && items[i].name != "Backstage passes to a TAFKAL80ETC concert" {
-			if items[i].quality > 0 {
-				if items[i].name != "Sulfuras, Hand of Ragnaros" {
-					items[i].quality = items[i].quality - 1
-				}
-			}
-		} else {
-			if items[i].quality < 50 {
-				items[i].quality = items[i].quality + 1
-				if items[i].name == "Backstage passes to a TAFKAL80ETC concert" {
-					if items[i].sellIn < 11 {
-						if items[i].quality < 50 {
-							items[i].quality = items[i].quality + 1
-						}
-					}
-					if items[i].sellIn < 6 {
-						if items[i].quality < 50 {
-							items[i].quality = items[i].quality + 1
-						}
-					}
-				}
-			}
-		}
+// simple func to return a max of 2 integers
+// useful in this context to stop items value going under 0
+func max(x int, y int) int {
+	if x == y {
+		return x
+	}
+	if x > y {
+		return x
+	}
+	return y
+}
 
-		if items[i].name != "Sulfuras, Hand of Ragnaros" {
-			items[i].sellIn = items[i].sellIn - 1
-		}
+// UpdateQualityBrie handles updates for aged brie
+func UpdateQualityBrie(item *Item) {
+	// always decrement sellInn
+	defer func() {
+		item.sellIn--
+	}()
+
+	// quality cannot go above 50
+	if item.quality == 50 {
+		return
+	}
+
+	// increment when prior to sellIn
+	if item.sellIn > 0 {
+		item.quality++
+	}
+	// increment by 2 if past sellInn date
+	if item.sellIn <= 0 {
+		item.quality = min(item.quality+2, 50)
+	}
+
+}
+
+// UpdateQualityBackstagePasses handles updates for backstage passes
+func UpdateQualityBackstagePasses(item *Item) {
```

Comment:
I like the idea of moving logic for each item to the separate function, it makes the code easier to follow and to add new items to the shop inventory.